### ISSUE TITLE
Script 2 - error when downloading elevation for study area using elevatr

### ISSUE THREAD

**Issue by ShipleyWill** [Hussein-Mahfouz/cycle-networks#10]
I still get this error when running this line from script 2:

`# download elevation for study area using elevatr
city_elev <- elevatr::get_elev_raster(city_geom, z= 9, prj = prj_elev, expand = 0.05)`

The error reads:
`Error in if ((is.null(sp::wkt(locations)) | nchar(sp::wkt(locations)) ==  : 
  argument is of length zero
In addition: Warning messages:
1: In sp::wkt(locations) : CRS object has no comment
2: In sp::wkt(locations) : CRS object has no comment
3: In sp::wkt(locations) : CRS object has no comment`

I'll press on and see what happens.

**Comment by ShipleyWill** [comment]
No luck. Next line (when ignoring the previous error) also returns an error

`> raster::writeRaster(city_elev, paste0("data/", chosen_city, '/city_elev.tif'))
Error in h(simpleError(msg, call)) : 
  error in evaluating the argument 'x' in selecting a method for function 'writeRaster': object 'city_elev' not found`

**Comment by Robinlovelace** [comment]
This may be a genuine issue associated with updates in the R world. Can you reproduce the issue with the latest versions of packages installed @Hussein-Mahfouz ?

**Comment by Hussein-Mahfouz** [comment]
@ShipleyWill I updated the packages but couldn't reproduce the error. Take a look here:
```
# download elevation for study area using elevatr
city_elev <- elevatr::get_elev_raster(city_geom, z= 9, prj = prj_elev, expand = 0.05)
```
Notice that I am using` elevatr::get_elev_raster`. This tells R to load the elevatr package. (It will only work if you have `elevatr` installed). Probably this is bad practice and I should have written library(elevatr) in the beginning. Could you try:

1) Make sure to install the following packages `raster` and `elevatr`
2) Make sure that the variable `city_geom` is not an empty sf (i.e. it does not have 0 rows. It could have 0 rows if Bradford is spelled incorrectly (spelling is case sensitive. You need a capital B)
3) Change the following line by adding an overwrite argument:
```
# save
raster::writeRaster(city_elev, paste0("data/", chosen_city, '/city_elev.tif'), overwrite = TRUE)
```
This will prevent an error being raised if you are trying to rerun the script and save a file that already exists (it overwrites that file). I will push this update from my side, but for now you can edit it yourself

Let us know if this works

**Comment by ShipleyWill** [comment]
Thanks for all this - here's how I got on:

1. I already have both of these packages installed, but I re-installed them anyway.
2. When I view `city_geom` I get a matrix with 39 rows and 7 columns - all looks promising
3. I edited the script as you instructed, and the script ran to the end, but I got several errors and warnings about various objects not found. The output in the 'Plots' folder was a blank .png file.

4. I deleted everything in the 'Data' folder, and started again with script 1. Before running script 1, I noticed this line:
`##### CHOOSE YOU CITY `
`if(!exists("chosen_city")) chosen_city <- "Manchester"`
**I had previously assumed that by entering `chosen_city <- "Bradford"` before running any scripts, I wouldn't need to edit any of the scripts. Re-reading the instructions, I think I was wrong! Is this line in script 1 changing my chosen city to Manchester? Is this my mistake?**

5. I edited script one from Manchester to Bradford. This worked fine. I got a couple of warnings like this:
ggrepel: 3 unlabeled data points (too many overlaps). Consider increasing max.overlaps 
Which I ignored.
6. Running Script 2 again now works! I get several errors and warnings, but there is an output in the 'Plots' folder (attached).
![boxplot_weighted_unweighted_distances](https://user-images.githubusercontent.com/86323018/123264934-6cf37200-d4f2-11eb-9084-9b4e1fac0234.png)

**Comment by ShipleyWill** [comment]
Here are the errors from Script 2. Can I continue?

`> # download elevation for study area using elevatr`
`> city_elev <- elevatr::get_elev_raster(city_geom, z= 9, prj = prj_elev, expand = 0.05)`
```
Error in if ((is.null(sp::wkt(locations)) | nchar(sp::wkt(locations)) ==  : 
  argument is of length zero
```


`> # save`
`> raster::writeRaster(city_elev, paste0("data/", chosen_city, '/city_elev.tif'), overwrite = TRUE)`
```
Error in h(simpleError(msg, call)) : 
  error in evaluating the argument 'x' in selecting a method for function 'writeRaster': object 'city_elev' not found
```


`> streetnet_sc <- osmdata::osm_elevation(streetnet_sc, elev_file = paste0("data/", chosen_city, '/city_elev.tif'))`
```
Error in check_elev_file(elev_file) : 
  file data/Bradford/city_elev.tif does not exist
```


`> # get routes column then slope column (function 1 then function 2)`
`> #flows_slope <- flows_slope %>% route(net = net) %>% slope(elev = uk_elev)`
`> flows_slope <- flows_slope %>% route(net = net) %>% slope(elev = city_elev)`
```
Error in h(simpleError(msg, call)) : 
  error in evaluating the argument 'x' in selecting a method for function 'extract': object 'city_elev' not found
In addition: Warning message:
In NextMethod() :
```


`> flows_slope %>% `
`+     dplyr::select(-c(cent_orig, cent_dest, route)) %>%`
`+     write_csv(path = paste0("data/",chosen_city,"/flows_dist_elev_for_potential_flow.csv"))`
```
Error: Can't subset columns that don't exist.
x Column `route` doesn't exist.
```


And some warnings:
```
Warning messages:
1: In rm(city_geom, city_names, dist_mat, dist_mat_1, dist_mat_2, dist_mat_3,  :
  object 'uk_elev' not found
2: In rm(city_geom, city_names, dist_mat, dist_mat_1, dist_mat_2, dist_mat_3,  :
  object 'city_elev' not found
```

**Comment by Robinlovelace** [comment]
I think you need the object `city_elev` which is not being created. It seems the code that creates that object is failing. I'm not sure but this may be due to the most recent version of the `sf` package. Keep chipping away, let's keep trying different things, and I'm sure we'll get there (can try myself next week but busy this week).

**Comment by Hussein-Mahfouz** [comment]
@ShipleyWill you could try one more thing while we try to figure out why it isn't working:

The line of code that is failing is an attempt to get raster data for Bradford through the elevatr package. You will see from line 128 that this is **Option 2**. You could try something similar to the old **Option 1** which will require you to do a preliminary manual step:

Step 1: 
- Run script 2 up to and including line 112

Step 2: 
- go to https://remotepixel.ca/projects/srtm1arc-gl.html and select the tile / tiles that cover Bradford. 
- click on create and input your email. They will send you a .tif file

Step 3: 
- Save the .tif file as **city_elev.tif** in this path 'data/Bradford/[add file here]'

Step 4:
- Read in the elevation data
```
city_elev <- raster::raster(paste0("data/", chosen_city, '/city_elev.tif'))
# specify the crs
prj_elev <- "+init=EPSG:4326"
# reproject
city_elev <- city_elev %>%
raster::projectRaster(., crs=prj_elev)
```

Step 5: 
- Jump to this line of code 
```
streetnet_sc <- osmdata::osm_elevation(streetnet_sc, elev_file = paste0("data/", chosen_city, '/city_elev.tif'))
```
Basically you can skip all code between line 112 and line 137

Let us know if this works

**Comment by ShipleyWill** [comment]
Thanks - it has worked!

No errors, one warning:
```
Warning message:
In rm(city_geom, city_names, dist_mat, dist_mat_1, dist_mat_2, dist_mat_3,  :
  object 'uk_elev' not found
```

**Comment by ShipleyWill** [comment]
The rest of the scripts have all competed - thanks for guiding me through this!

I'll email the plots to Robin (Hussein - I don't have your email) - in case there is anything obvious that I've done wrong.

If there's anything I can do for you, please let me know.

Now off to interpret these plots...

**Comment by Robinlovelace** [comment]
Great to hear you got it working. @ShipleyWill it would be great if you send a tweet spreading the word that you've reproduced the results and encouraging others to do so, maybe with an image of the results for Bradford (I'm interested in how they'll come out also).

Impressive work, I gather you're an engineer so training in technical problem solving came in handy I guess!

**Comment by ShipleyWill** [comment]
I will do. Thanks again for all your help. I will bring it up at our next Living Streets meeting - there are 50 groups in different locations across the UK, so I’m sure somebody else will want to take it for a spin.

I’m only pretending to be an engineer (I’m a biologist really) although slightly more convincingly than I’m pretending to be a transport modeller!


________________________________

**Comment by Robinlovelace** [comment]
Great, well I deem this issue closed. Will be great if you can get others interested in the methods/results, any feedback on that, keep us posted!
